### PR TITLE
feat(prisma-next): source EQL v3 install SQL from @cipherstash/eql at runtime

### PR DESCRIPTION
**Pull request by coderdan**
## Why

Baking the \~1.7 MB EQL v3 install bundle into the v3 baseline migration's `ops.json` meant every `@cipherstash/eql` upgrade required re-running the maintainer emit loop and committing another generated SQL blob. This is the concrete cost @coderdan flagged.

This is the "depend, don't bake" direction from cipherstash/stack#690 (PR3), scoped to what actually matters: the sourcing model. It does **not** change the contract-space head-ref/invariants or require framework-migration-emitter changes, because the SQL is not part of the contract storage hash.

## What changed

* `@cipherstash/eql` is now an **exact-pinned runtime dependency** (`3.0.0`, previously a devDependency), matching the EQL version used by `@cipherstash/stack` and the `stash` CLI.
* The committed v3 baseline op carries a **sentinel placeholder** instead of baked SQL. `control.ts` uses `withRuntimeEqlSqlPackage()` to inject `readInstallSql()` from the installed `@cipherstash/eql` and recompute the migration hash from the resulting operations.
* Upgrading the coordinated EQL pin now requires a dependency bump and rebuild/release, but **no 1.7 MB migration re-emit**.
* `ops.json`: **1,717,297 → 1,219 bytes.**

## Why it's safe (no framework surgery)

The v3 baseline is an **invariant-only self-edge** (`from === to`; the bundle declares no contract-space storage). Therefore:

* the install SQL does not change the contract-space storage hash;
* runtime insertion recomputes the content-addressed `migrationHash` from the injected operations before Prisma Next materializes the package;
* Prisma Next's canonical disk reader subsequently verifies that hash, preserving the normal migration-integrity check;
* the apply planner routes on the `cipherstash:install-eql-v3-bundle-v1` invariant, not SQL content.

Verified against `@prisma-next/migration-tools@0.14.0`.

## Trusted SQL bundle insertion (not user-input SQL injection)

Here, “inject” means inserting a trusted package asset into the migration descriptor. It is not a path for arbitrary or user-provided SQL:

* SQL comes only from exact-pinned `@cipherstash/eql@3.0.0`; no user input is interpolated.
* A unique sentinel identifies the replacement location, and a missing sentinel fails immediately.
* Only the sentinel step's `sql` field is replaced; sibling steps and migration fields are preserved.
* The operation class is validated and the canonical migration hash is recomputed after replacement.
* A descriptor → disk → integrity-checked read round trip verifies the materialized package.
* The pnpm lockfile pins the npm artifact's registry integrity.

The trust chain is: **exact-pinned package → trusted static SQL asset → exact sentinel replacement → recomputed content hash → verified disk round trip**.

This does not sanitize arbitrary SQL or make user-generated raw SQL safe; it prevents untrusted input from entering this path.

## User-facing impact

**None.** EQL still installs as part of `prisma-next migration apply`. The v2 baseline is untouched (still baked; v2 is legacy/unreleased).

## Tests

* The committed `ops.json` carries the sentinel and remains under 5 KB.
* The descriptor injects `readInstallSql()` and recomputes the runtime migration hash.
* A regression test materializes the descriptor package and reads it back through Prisma Next's integrity-checking migration reader.
* The live-PG migration test compares the applied SQL with the runtime descriptor operation.
* 602 Prisma Next tests pass; 30 credentialed live tests skip without credentials/Postgres. Build, typecheck, and `code:check` pass.

## For @calvinbrewer

This changes the v3 baseline authoring model: `migration.ts` emits a placeholder, while the runtime descriptor injects the SQL and recomputes the content-addressed migration hash. The re-emit loop (`pnpm exec tsx migrations/.../migration.ts`) still works and is a deterministic no-op. Flagging since it touches the contract-space baseline you own—the invariant/head-ref wiring is unchanged.

Part of cipherstash/stack#690 (the "depend on @cipherstash/eql, don't bake" thread).

[https://claude.ai/code/session_01MxTTPaPP16m6br7Hoab94w](<https://claude.ai/code/session_01MxTTPaPP16m6br7Hoab94w>)

## Summary by CodeRabbit

* **New Features**
  * EQL v3 installation SQL is now resolved via the installed EQL package during migration/descriptor build-time injection.
  * Baseline migration artifacts no longer embed the full v3 install SQL bundle.
* **Bug Fixes**
  * Migration packaging now injects and validates the install SQL consistently, updating migration hashes to match the materialized operations.
  * `@cipherstash/eql` is now a runtime dependency to stay aligned with v3 domain expectations.
* **Tests**
  * Updated and expanded v3 and live migration-apply coverage to verify placeholder usage, runtime SQL resolution, and drift-guard replacement behavior.

### CODE REVIEW

File: packages/prisma-next/test/v3/migration-v3.test.ts [hunk 1]
```diff
@@ -39,16 +41,34 @@ describe('v3 baseline migration (20260601T0100_install_eql_v3_bundle)', () => {
...
       }>
     )[0]!
-    expect(op.execute?.[0]?.sql).toBe(readInstallSql())
+    // The ~1.7 MB bundle must not be committed here — an EQL patch/minor should
+    // not require re-emitting this file. The op carries the sentinel instead.
+    expect(op.execute?.[0]?.sql).toBe(RUNTIME_EQL_SQL_SENTINEL)
+    expect(op.execute?.[0]?.sql).not.toContain('CREATE')
+    expect(JSON.stringify(v3Ops).length).toBeLessThan(5_000)
     expect(releaseManifest.eqlVersion).toBe('3.0.0')
   })
```
[the remark this answers is not in the document]
Comment (coderdan) [reply packages/prisma-next/test/v3/migration-v3.test.ts:78]: False positive — there is no duplicate `execute` here. Both type assertions declare `execute?` exactly once (line 50 in the first test, line 70 in the second — the second also declares `id`, no repeat). There is no `TS2300`.

This looks inferred from the failed pipeline ("Source: Pipeline failures") rather than the current code. The actual typecheck failures were `TS2459` (`readInstallSql` was dropped from the `eql-bundle-v3` re-export the live-test helpers import) and `TS18048` (`contractSpace` possibly undefined) — both fixed in 95d7f39d. `pnpm --filter @cipherstash/prisma-next run typecheck` now exits 0 and CI Node 22/24 are green. Resolving.

**Comment by coderdan** [comment]
# Differential Review: PR 694

Codex/gpt-5.6-sol

## Executive Summary

| Severity | Count |
|---|---:|
| 🔴 CRITICAL | 0 |
| 🟠 HIGH | 1 |
| 🟡 MEDIUM | 1 |
| 🟢 LOW | 0 |

**Overall Risk:** HIGH  
**Recommendation:** REJECT until the migration hash is recomputed after runtime SQL injection and the fresh-seed round trip is tested.

**Key Metrics:**

- Files analyzed: 9/9 (100%)
- Changed lines: +214 / -40
- High-risk production entry points changed: 1 (`@cipherstash/prisma-next/control`)
- Test gaps: 1 fresh descriptor → disk → verified-read workflow
- Security regressions: 0 identified

## What Changed

**Commit Range:** `08113301b524003fb2906240115bec644fccf069..95d7f39d8a3179adb7421bf15d77c7073d10c474`  
**Commits:** 3  
**Timeline:** 2026-07-18

| File group | Risk | Summary |
|---|---|---|
| `src/exports/control.ts`, `src/migration/eql-bundle-v3.ts` | HIGH | Replaces the committed sentinel with installed EQL SQL at descriptor import time |
| v3 migration JSON/TS | HIGH | Replaces baked SQL with a sentinel and changes the content hash |
| `package.json`, `pnpm-lock.yaml` | MEDIUM | Moves exact `@cipherstash/eql@3.0.0` from dev to runtime dependencies |
| tests | MEDIUM | Covers in-memory injection, but not materialization and verified reload |
| changeset | LOW | Documents the published behavior change |

## Critical Findings

### 🟠 HIGH: Freshly seeded migrations fail their content-addressed integrity check

**File:** `packages/prisma-next/src/exports/control.ts:89`  
**Commit:** `e566f93339bfcc89ed583470477a02e96578ebf3`  
**Blast Radius:** Every fresh Prisma Next project that seeds this extension migration  
**Test Coverage:** NO — current tests stop at the in-memory descriptor

**Description:**

`control.ts` keeps `v3BaselineMetadata` unchanged while replacing the sentinel in `v3BaselineOps` with the 1.7 MB runtime SQL. Prisma Next's seed phase materializes both descriptor values into the user's migration directory. The next load recomputes `migrationHash` over the mutated operations and rejects the package because the stored hash was computed over the sentinel.

Observed values:

```text
stored metadata hash:    sha256:55f261c28b9fe8feee21270a652d758a5c90ccbe7aac9cc79ac50662b956160b
hash after SQL injection: sha256:d1ca5a987ab10eb85c2fe05700134b31f17f694d2d5b3df3969e902610b1d9cf
read result: MIGRATION.HASH_MISMATCH
```

**Historical Context:**

- The v3 baseline and its content-addressed migration were introduced in `a75513b45279bd9bddc8da1bae417a811974b280`.
- The baseline deliberately baked the SQL so `migration.json` and `ops.json` had one content identity.
- This PR changes the operations after that identity has been computed.

**Failure Scenario:**

1. A fresh user imports `@cipherstash/prisma-next/control`.
2. `withRuntimeEqlSql()` injects the installed EQL SQL into the descriptor operations.
3. `migration plan` seeds the descriptor migration into `migrations/cipherstash/...`.
4. Phase 2 reads the newly written package and calls `verifyMigrationHash()`.
5. The command stops with `MIGRATION.HASH_MISMATCH`; EQL and the application migration are not applied.

**Proof of Concept:**

```js
const pkg = descriptor.contractSpace.migrations[1]
await materialiseMigrationPackage(root, pkg)
await readMigrationPackage(join(root, pkg.dirName))
// MigrationToolsError: MIGRATION.HASH_MISMATCH
```

The repository's Prisma example E2E does not expose this because its migration directory already contains the old baked package; the seed phase skips existing directories by design.

**Recommendation:**

Build the runtime migration package atomically: inject the SQL and recompute `metadata.migrationHash` with `computeMigrationHash(metadata, injectedOps)` before passing it to `contractSpaceFromJson`. Add a test that materializes the descriptor package into an empty directory and successfully reads it back with `readMigrationPackage`.

### 🟡 MEDIUM: The credentialed live migration test still asserts the removed baked-SQL model

**File:** `packages/prisma-next/test/live/migration-apply-live-pg.test.ts:82`  
**Commit:** unchanged from baseline, invalidated by `e566f93339bfcc89ed583470477a02e96578ebf3`  
**Blast Radius:** `pnpm --filter @cipherstash/prisma-next test:live` when live credentials and `DATABASE_URL` are present  
**Test Coverage:** The default run skips this suite

**Description:**

The test reads committed `ops.json` directly and asserts `op.execute[0].sql === readInstallSql()`. The committed value is now the 124-character sentinel, while `readInstallSql()` is the full bundle, so this assertion is guaranteed to fail once the live gate opens. Its header and test name also still describe SQL as baked into `ops.json`.

**Recommendation:**

Read the operation from the injected descriptor (or, preferably, from the materialized-and-verified package used by the real CLI path), then apply and compare that SQL. Update the stale comments and test name.

## Test Coverage Analysis

Local results:

- ✅ `pnpm --filter @cipherstash/prisma-next build`
- ✅ `pnpm --filter @cipherstash/prisma-next typecheck`
- ✅ default package tests: 601 passed, 30 live tests skipped
- ✅ `pnpm run code:check` (warnings only, as allowed by repository policy)
- ❌ descriptor migration materialize/read reproduction: `MIGRATION.HASH_MISMATCH`
- ⚠️ live-PG suite could not execute without credentials/database; static comparison proves its SQL identity assertion is false (`124` sentinel characters versus `1,666,263` runtime SQL characters)

The highest-value invariant test is a round trip: descriptor package → materialize → verified read. That property is currently absent.

## Blast Radius Analysis

| Changed function/path | Direct production callers | Effective impact | Priority |
|---|---:|---|---|
| `withRuntimeEqlSql()` | 1 (`control.ts`) | All fresh consumers of the public control descriptor | P1 |
| `@cipherstash/eql` runtime SQL loader | 1 descriptor import path | Control-plane startup and migration authoring | P1 |

The repository contains three direct consumers of `cipherstashExtensionDescriptor`; external consumers of the published `./control` subpath are not enumerable locally.

## Historical Context

No authorization, crypto primitive, plaintext-logging, or payload-shape guard was removed. The relevant removed behavior was content identity: commit `a75513b4` emitted the exact SQL that its metadata hash attested. The PR preserves the SQL bytes but mutates them after loading the attestation.

The exact `@cipherstash/eql@3.0.0` pin matches `@cipherstash/stack` and the CLI, and the lockfile retains registry integrity metadata. No additional dependency version was introduced; GitHub's OSV checks are green.

## Recommendations

### Immediate (Blocking)

- [x] Recompute the v3 migration hash from the injected operations.
- [x] Add a fresh-seed materialize/read test using Prisma Next's real migration I/O helpers.
- [x] Fix the stale live-PG migration test to exercise the injected/materialized operation.

### Before Production

- [ ] Add a fresh-project Prisma Next E2E fixture; the current checked-in fixture only tests the existing-directory path.

### Technical Debt

- [x] Consider representing runtime injection as a helper returning `{ metadata, ops }` so callers cannot update one half of the content-addressed package independently.

## Analysis Methodology

**Strategy:** SURGICAL (751 relevant source/config files; 9 changed files)

**Coverage:**

- Changed files reviewed: 9/9
- High-risk diff regions: 100%
- One-hop framework dependencies traced through descriptor seeding, materialization, hash verification, and aggregate loading
- Baseline history and removed behavior reviewed with `git log`, `git blame`, and pickaxe search
- Adversarial model: an ordinary fresh-project user, with no special privileges, triggers a deterministic availability failure through the documented migration workflow

**Limitations:**

- No live ZeroKMS/Postgres credentials were available locally.
- External consumers and future Prisma Next versions were not analyzed.

**Confidence:** HIGH

## Appendix: Commit References

| Commit | Purpose |
|---|---|
| `a75513b4` | Baseline EQL v3 migration and live tests |
| `e566f933` | Runtime SQL sourcing |
| `12094257` | Exact pin, sentinel matching, fail-fast injection |
| `95d7f39d` | Re-export/typecheck follow-up |

**Comment by coderdan** [comment]
Addressed the review findings in `01f1d15a`.

- Runtime EQL SQL injection now recomputes the content-addressed migration hash atomically with the injected ops.
- Added a descriptor → disk → integrity-checked read regression test using Prisma Next’s canonical migration I/O helpers.
- Updated the live-PG migration test to compare against the runtime descriptor op instead of the committed sentinel fixture.
- Updated the migration documentation and changeset.

Validation:
- `pnpm --filter @cipherstash/prisma-next build`
- `pnpm --filter @cipherstash/prisma-next typecheck`
- `pnpm --filter @cipherstash/prisma-next test` — 602 passed, 30 credentialed live tests skipped
- `pnpm run code:check`
- Fresh-seed reproduction now materializes and verifies successfully with hash `sha256:d1ca5a987ab10eb85c2fe05700134b31f17f694d2d5b3df3969e902610b1d9cf`.

**Comment by coderdan** [comment]
Comprehensive human and AI review (Coderabbit and Codex/gpt-5.6-sol)
